### PR TITLE
drivers: timer: implementation cleanups

### PR DESCRIPTION
**Pull request by gmarull**
This PR started as a discussion to remove sys_clock* weak implementations, but I've also taken the opportunity to clean up a bit of other areas, in particular:

- Kconfig split
- Move init to every driver (no need for weak init/PM functions)
- Improve usage of sys_clock_disable() API
- Improve usage of SysTick sys_clock_isr() install

A couple of weak functions remain `sys_clock_elapsed` and `sys_clock_set_timeout`. By looking at the implementations and docs, I think we can also do better in that area, but I may be missing things:

- Simplify `sys_clock_elapsed()` (provide a default that returns 0 if non-tickless?)
- Simplify `sys_clock_set_timeout()` (no-op if non-tickless?)

### CODE REVIEW

**Comment by cfriedt** [comment]
LGTM - I think the main purpose of weak global functions is to have a sane fallback when a more specific implementation does not exist. They also override at link time. It's possible to do the same thing at runtime but that does come with a small footprint penalty.

If the system clock is a singleton device such that an implementation must exist and no other implementation may exist, then I think the approach here makes sense.

**Review by cfriedt — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by andyross — COMMENTED** [review]
Can I be a bit cranky here?  This obviously isn't a big deal (the weak linkage was originally introduced to support a bunch of drivers being ported over from a previous incarnation of the driver interface), but it seems like this is happening just to paper over a bug with weak linkage on OS X?  Maybe that's something we should pay more attention to?  This isn't the only spot in Zephyr where we're using default linkage and it probably won't be the last.

Mostly seems like this is a slightly kludgey workaround to a problem whose root cause we really want to fix instead.

**Comment by gmarull** [comment]
> Can I be a bit cranky here? This obviously isn't a big deal (the weak linkage was originally introduced to support a bunch of drivers being ported over from a previous incarnation of the driver interface), but it seems like this is happening just to paper over a bug with weak linkage on OS X? Maybe that's something we should pay more attention to? This isn't the only spot in Zephyr where we're using default linkage and it probably won't be the last.
> 
> Mostly seems like this is a slightly kludgey workaround to a problem whose root cause we really want to fix instead.

I really saw the OS X issue after opening this proposal, so the main purpose was not to fix that problem but discuss if weak symbols make sense in this case.

**Comment by cfriedt** [comment]
> Can I be a bit cranky here? This obviously isn't a big deal (the weak linkage was originally introduced to support a bunch of drivers being ported over from a previous incarnation of the driver interface), but it seems like this is happening just to paper over a bug with weak linkage on OS X? Maybe that's something we should pay more attention to? This isn't the only spot in Zephyr where we're using default linkage and it probably won't be the last.
> 
> Mostly seems like this is a slightly kludgey workaround to a problem whose root cause we really want to fix instead.

I agree for the most part.

With macOS, weak linking (in the ELF sense) simply was never supported - it's not that it's buggy. macOS supports a different weak construct to do with frameworks that is not exactly the same.

I've experienced issues with weak linkage and various vendor toolchains - e.g. IAR (it's been a while, thank the skies). Default link ordering makes weak linking fragile as well.

Beyond all of that, the main reason I'm not a fan of weak linking is because it's not part of ISO C, which is (at least in theory) mostly portable. I would say that Kconfig could probably be used to solve a more general problem that weak linking solves, which is also a build-time solution.

**Comment by gmarull** [comment]
@cfriedt @carlescufi can you re-review?

**Review by cfriedt — COMMENTED** [review]
@gmarull - Looks good to me overall.

I like that you split out Kconfig into separate files. Inheriting copyright seems a bit much, but I'm not a lawyer.

Also good thinking about using `SYS_INIT()` instead of global `__weak` symbols.

It seems like a number of large blocks of C code were moved. Was that necessary in every case? Would converting to static work while keeping the code in its original place?

**Comment by gmarull** [comment]
> It seems like a number of large blocks of C code were moved. Was that necessary in every case? Would converting to static work while keeping the code in its original place?

I did that since init function is usually placed at the bottom, but I can drop that change if preferred.

**Review by dleach02 — APPROVED** [review]

**Comment by gmarull** [comment]
@nashif @carlescufi This PR should now be ready, the 2 test failures are due to https://github.com/zephyrproject-rtos/zephyr/issues/39978

**Comment by gmarull** [comment]
@andyross can you please review this one?

**Review by carlescufi — APPROVED** [review]
Looks good to me, thanks! We need @andyross to review however, before we merge.

**Review by stephanosio — APPROVED** [review]

**Review by andyross — APPROVED** [review]
Sorry, I missed this when it was submitted and had to be pointed at it.  A few nitpicks if someone wants to take time to accomodate, but nothing that should block merge.

File: drivers/timer/Kconfig
Comment (gmarull) [reply drivers/timer/Kconfig]: changed to `SYSTEM_TIMER_HAS_DISABLE_SUPPORT`

File: drivers/timer/Kconfig
Comment (gmarull) [reply drivers/timer/Kconfig:58]: I think the majority of drivers have split Kconfig files, e.g. adc, counter, gpio... It is of course subjective, but I find it more readable to have separate Kconfig files. Not all drivers are boolean: stm32, cortex-m systick, nrf rtc, etc.

**Pull request merged**
